File: evidence-service/tests/session-gate.test.ts
/**
 * Session Gate Tests
 *
 * Proves:
 * 1. Regular session hours permit acquisition
 * 2. Weekends block acquisition
 * 3. Full holidays block acquisition
 * 4. Early-close days block acquisition after 13:30 ET
 * 5. Early-close days permit acquisition before 13:30 ET
 * 6. Pre-market blocks acquisition
 * 7. Post-market blocks acquisition on standard days
 */

import { describe, it, expect } from "vitest";
import { isAcquisitionPermitted } from "../src/acquisition-worker.js";

/** Create a Date at a specific ET time on a given date.
 *  Uses the same DST approximation as the session gate:
 *  EDT (UTC-4) for months Apr–Oct, EST (UTC-5) otherwise. */
function etDate(dateStr: string, hours: number, minutes: number): Date {
  const [year, month, day] = dateStr.split("-").map(Number);
  // Match the gate's DST logic: (month > 3 && month < 11) → EDT, else EST
  const isEDT = (month > 3 && month < 11) || (month === 3 && day >= 8) || (month === 11 && day < 1);
  const utcOffset = isEDT ? 4 : 5; // hours to add to ET to get UTC
  return new Date(Date.UTC(year, month - 1, day, hours + utcOffset, minutes, 0));
}

describe("session gate — regular session", () => {
  it("permits acquisition at 10:00 ET on a trading day", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-15", 10, 0)); // Wednesday
    expect(result.permitted).toBe(true);
    expect(result.reason).toBe("Regular session");
  });

  it("permits acquisition at 15:00 ET on a trading day", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-15", 15, 0));
    expect(result.permitted).toBe(true);
  });

  it("permits acquisition at 16:14 ET (within delay drain)", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-15", 16, 14));
    expect(result.permitted).toBe(true);
  });
});

describe("session gate — weekends", () => {
  it("blocks acquisition on Saturday", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-18", 12, 0)); // Saturday
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Weekend");
  });

  it("blocks acquisition on Sunday", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-19", 12, 0)); // Sunday
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Weekend");
  });
});

describe("session gate — holidays", () => {
  it("blocks acquisition on July 4 observed (July 3, 2026)", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-03", 12, 0)); // Thursday
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Exchange holiday");
  });

  it("blocks acquisition on Thanksgiving", () => {
    const result = isAcquisitionPermitted(etDate("2026-11-26", 12, 0)); // Thursday
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Exchange holiday");
  });
});

describe("session gate — early-close days", () => {
  it("permits acquisition at 12:00 ET on day after Thanksgiving", () => {
    const result = isAcquisitionPermitted(etDate("2026-11-27", 12, 0)); // Friday
    expect(result.permitted).toBe(true);
    expect(result.reason).toBe("Regular session");
  });

  it("permits acquisition at 13:29 ET on day after Thanksgiving (within early window)", () => {
    const result = isAcquisitionPermitted(etDate("2026-11-27", 13, 29));
    expect(result.permitted).toBe(true);
  });

  it("blocks acquisition at 13:31 ET on day after Thanksgiving (past early close + delay)", () => {
    const result = isAcquisitionPermitted(etDate("2026-11-27", 13, 31));
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Early close");
  });

  it("blocks acquisition at 14:00 ET on day after Thanksgiving", () => {
    const result = isAcquisitionPermitted(etDate("2026-11-27", 14, 0));
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Early close");
  });

  it("permits acquisition at 10:00 ET on Christmas Eve", () => {
    const result = isAcquisitionPermitted(etDate("2026-12-24", 10, 0)); // Thursday
    expect(result.permitted).toBe(true);
  });

  it("blocks acquisition at 14:00 ET on Christmas Eve", () => {
    const result = isAcquisitionPermitted(etDate("2026-12-24", 14, 0));
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Early close");
  });
});

describe("session gate — off-hours", () => {
  it("blocks acquisition at 09:00 ET (pre-market)", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-15", 9, 0));
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Pre-market");
  });

  it("blocks acquisition at 16:30 ET (post-market)", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-15", 16, 30));
    expect(result.permitted).toBe(false);
    expect(result.reason).toContain("Market closed");
  });

  it("blocks acquisition at 20:00 ET (evening)", () => {
    const result = isAcquisitionPermitted(etDate("2026-07-15", 20, 0));
    expect(result.permitted).toBe(false);
  });
});

File: evidence-service-java/src/test/java/com/wheelwright/evidence/SessionGateTest.java
package com.wheelwright.evidence;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Session Gate Tests — fixed instants mirroring TypeScript session-gate.test.ts.
 *
 * Uses the same DST approximation as the TypeScript implementation:
 *   EDT (UTC-4) for summer months, EST (UTC-5) otherwise.
 */
class SessionGateTest {

    private final SessionGate gate = new SessionGate(Clock.systemUTC());

    /**
     * Create an Instant at a specific ET time on a given date.
     * Matches the TypeScript test helper: adds UTC offset to ET hours.
     */
    private static Instant etInstant(String dateStr, int etHours, int etMinutes) {
        LocalDate date = LocalDate.parse(dateStr);
        int month = date.getMonthValue();
        int day = date.getDayOfMonth();
        // Same DST logic as TypeScript gate
        boolean isEDT = (month > 3 && month < 11) || (month == 3 && day >= 8) || (month == 11 && day < 1);
        int utcOffset = isEDT ? 4 : 5; // hours to add to ET to get UTC
        int utcHours = etHours + utcOffset;
        // Handle day rollover (e.g., 20:00 ET + 4 = 24:00 UTC = next day 00:00)
        LocalDate utcDate = date;
        if (utcHours >= 24) {
            utcHours -= 24;
            utcDate = date.plusDays(1);
        }
        return ZonedDateTime.of(utcDate, LocalTime.of(utcHours, etMinutes),
            ZoneOffset.UTC).toInstant();
    }

    @Nested
    @DisplayName("regular session")
    class RegularSession {

        @Test
        @DisplayName("permits acquisition at 10:00 ET on a trading day")
        void permitsAt1000ET() {
            var result = gate.isPermitted(etInstant("2026-07-15", 10, 0)); // Wednesday
            assertTrue(result.permitted());
            assertEquals("Regular session", result.reason());
        }

        @Test
        @DisplayName("permits acquisition at 15:00 ET on a trading day")
        void permitsAt1500ET() {
            var result = gate.isPermitted(etInstant("2026-07-15", 15, 0));
            assertTrue(result.permitted());
        }

        @Test
        @DisplayName("permits acquisition at 16:14 ET (within delay drain)")
        void permitsAt1614ET() {
            var result = gate.isPermitted(etInstant("2026-07-15", 16, 14));
            assertTrue(result.permitted());
        }
    }

    @Nested
    @DisplayName("weekends")
    class Weekends {

        @Test
        @DisplayName("blocks acquisition on Saturday")
        void blocksSaturday() {
            var result = gate.isPermitted(etInstant("2026-07-18", 12, 0)); // Saturday
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Weekend"));
        }

        @Test
        @DisplayName("blocks acquisition on Sunday")
        void blocksSunday() {
            var result = gate.isPermitted(etInstant("2026-07-19", 12, 0)); // Sunday
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Weekend"));
        }
    }

    @Nested
    @DisplayName("holidays")
    class Holidays {

        @Test
        @DisplayName("blocks acquisition on July 4 observed (July 3, 2026)")
        void blocksJuly4Observed() {
            var result = gate.isPermitted(etInstant("2026-07-03", 12, 0)); // Thursday
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Exchange holiday"));
        }

        @Test
        @DisplayName("blocks acquisition on Thanksgiving")
        void blocksThanksgiving() {
            var result = gate.isPermitted(etInstant("2026-11-26", 12, 0)); // Thursday
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Exchange holiday"));
        }
    }

    @Nested
    @DisplayName("early-close days")
    class EarlyClose {

        @Test
        @DisplayName("permits acquisition at 12:00 ET on day after Thanksgiving")
        void permitsAt1200OnBlackFriday() {
            var result = gate.isPermitted(etInstant("2026-11-27", 12, 0)); // Friday
            assertTrue(result.permitted());
            assertEquals("Regular session", result.reason());
        }

        @Test
        @DisplayName("permits acquisition at 13:29 ET on day after Thanksgiving (within early window)")
        void permitsAt1329OnBlackFriday() {
            var result = gate.isPermitted(etInstant("2026-11-27", 13, 29));
            assertTrue(result.permitted());
        }

        @Test
        @DisplayName("blocks acquisition at 13:31 ET on day after Thanksgiving (past early close + delay)")
        void blocksAt1331OnBlackFriday() {
            var result = gate.isPermitted(etInstant("2026-11-27", 13, 31));
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Early close"));
        }

        @Test
        @DisplayName("blocks acquisition at 14:00 ET on day after Thanksgiving")
        void blocksAt1400OnBlackFriday() {
            var result = gate.isPermitted(etInstant("2026-11-27", 14, 0));
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Early close"));
        }

        @Test
        @DisplayName("permits acquisition at 10:00 ET on Christmas Eve")
        void permitsAt1000OnChristmasEve() {
            var result = gate.isPermitted(etInstant("2026-12-24", 10, 0)); // Thursday
            assertTrue(result.permitted());
        }

        @Test
        @DisplayName("blocks acquisition at 14:00 ET on Christmas Eve")
        void blocksAt1400OnChristmasEve() {
            var result = gate.isPermitted(etInstant("2026-12-24", 14, 0));
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Early close"));
        }
    }

    @Nested
    @DisplayName("off-hours")
    class OffHours {

        @Test
        @DisplayName("blocks acquisition at 09:00 ET (pre-market)")
        void blocksPreMarket() {
            var result = gate.isPermitted(etInstant("2026-07-15", 9, 0));
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Pre-market"));
        }

        @Test
        @DisplayName("blocks acquisition at 16:30 ET (post-market)")
        void blocksPostMarket() {
            var result = gate.isPermitted(etInstant("2026-07-15", 16, 30));
            assertFalse(result.permitted());
            assertTrue(result.reason().contains("Market closed"));
        }

        @Test
        @DisplayName("blocks acquisition at 20:00 ET (evening)")
        void blocksEvening() {
            var result = gate.isPermitted(etInstant("2026-07-15", 20, 0));
            assertFalse(result.permitted());
        }
    }
}
